quantization/src/encoded_vectors_binary.rs

```diff
@@ -5,27 +5,121 @@ use crate::{
     VectorParameters,
 };
 use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
 use std::path::Path;
 
-type BitsStoreType = u128;
-
-const BITS_STORE_TYPE_SIZE: usize = std::mem::size_of::<BitsStoreType>() * 8;
-
-pub struct EncodedVectorsBin<TStorage: EncodedStorage> {
+pub struct EncodedVectorsBin<TBitsStoreType: BitsStoreType, TStorage: EncodedStorage> {
     encoded_vectors: TStorage,
     metadata: Metadata,
+    bits_store_type: PhantomData<TBitsStoreType>,
 }
 
-pub struct EncodedBinVector {
-    encoded_vector: Vec<BitsStoreType>,
+pub struct EncodedBinVector<TBitsStoreType: BitsStoreType> {
+    encoded_vector: Vec<TBitsStoreType>,
 }
 
 #[derive(Serialize, Deserialize)]
 struct Metadata {
     vector_parameters: VectorParameters,
 }
 
-impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
+pub trait BitsStoreType:
+    Default + Copy + Clone + core::ops::BitOrAssign + std::ops::Shl<usize, Output = Self>
+{
+    fn bits_count() -> usize;
+
+    fn one() -> Self;
+
+    fn count_ones(self) -> usize;
+
+    /// Xor vectors and return the number of bits set to 1
+    ///
+    /// Assume that `v1` and `v2` are aligned to `BITS_STORE_TYPE_SIZE` with both with zeros
+    /// So it does not affect the resulting number of bits set to 1
+    fn xor_popcnt(v1: &[Self], v2: &[Self]) -> usize;
+}
+
+impl BitsStoreType for u8 {
+    fn bits_count() -> usize {
+        8
+    }
+
+    fn one() -> Self {
+        1
+    }
+
+    fn count_ones(self) -> usize {
+        self.count_ones() as usize
+    }
+
+    fn xor_popcnt(v1: &[Self], v2: &[Self]) -> usize {
+        debug_assert!(v1.len() == v2.len());
+
+        #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+        if is_x86_feature_detected!("sse4.2") {
+            unsafe {
+                return impl_xor_popcnt_sse_uint8(v1.as_ptr(), v2.as_ptr(), v1.len() as u32)
+                    as usize;
+            }
+        }
+
+        let mut result = 0;
```

Comment:
There is no `neon` here because `u8` alignment is designed for multivectors which are short. But `neon` has popcnt for 64bit and 128bit values only